pandas/core/arrays/base.py

```diff
@@ -406,8 +408,25 @@ def argsort(self, ascending=True, kind='quicksort', *args, **kwargs):
         # 1. _values_for_argsort : construct the values passed to np.argsort
         # 2. argsort : total control over sorting.
         ascending = nv.validate_argsort_with_ascending(ascending, args, kwargs)
-        values = self._values_for_argsort()
-        result = np.argsort(values, kind=kind, **kwargs)
+        values, mask = self._values_for_argsort()
+
+        def permutation(mask):
+            # Return a permutation which maps the indices of the
+            # subarray without nan to the indices of the original array.
+            permu = np.arange(len(mask))
+            permu = permu[~mask]
+            return permu
+
+        if mask.any():
+            notmask = ~mask
+            notnull = np.argsort(values[notmask], kind=kind, **kwargs)
+            permu = permutation(mask)
```

Comment:
we might have some utilities in pandas/core/sorting.py which are helpful here (or as a possible future PR to pull common code out)

Reply:
I can't find the common code in pandas/core/sorting.py. Can we go ahead? I want to implement argmin for EA based on this PR.